### ISSUE TITLE
Access to ICARUS

### ISSUE THREAD

**Issue by m-hadjali** [TIBHannover/iconclass-classification#2]
Hi,
I’m very interested in this project. Would it be possible to get access to the dataset ICARUS?
Thank you!

**Comment by SpringsteinM** [comment]
Hello, sorry for the silence here. I sat down with my co-authors again, and she created a Zenodo record. https://zenodo.org/records/18980670

**Comment by m-hadjali** [comment]
Hello, thanks a lot for publishing the dataset!
Would it be possible to share the **BLIP descriptions** used (or the script/backbone you used), as well as the **ICAI train/val/test splits**?
I also sent you an email about this :)

**Comment by SpringsteinM** [comment]
I uploaded the requested files here: https://huggingface.co/springsteinm/iconclass_cat/tree/main

**Comment by m-hadjali** [comment]
Thank you! However, the BLIP description file appears to be incomplete.
Could you also clarify which version of BLIP was used? I also could not find any script to generate the descriptions (CLIP, BLIP, KW) in the repository.

**Comment by SpringsteinM** [comment]
Thanks for the info. I fix the file and uploaded it again. 

Here is the script for generate the text labels.

```python
import os
import sys
import re
import argparse
import json
import logging
import random

from transformers import Blip2Processor, Blip2Model, AutoProcessor, Blip2ForConditionalGeneration
from transformers import InstructBlipProcessor, InstructBlipForConditionalGeneration

# from otter.modeling_otter import OtterForConditionalGeneration
import transformers

import torch

from transformers import LlamaTokenizer, BitsAndBytesConfig
from iconclass_utils import iconclass_dataloader


def parse_args():
    parser = argparse.ArgumentParser(description="")

    parser.add_argument("-v", "--verbose", action="store_true", help="verbose output")
    parser.add_argument("--label_path", help="verbose output")
    parser.add_argument("--mapping_path", help="verbose output")
    parser.add_argument("--output_path", help="verbose output")
    parser.add_argument("--box_path", nargs="+")
    parser.add_argument("--image_path", help="verbose output")
    parser.add_argument("-q", "--query", help="verbose output")
    parser.add_argument("-a", "--annotation", help="verbose output")
    parser.add_argument("-e", "--exists_path", help="verbose output")

    parser.add_argument("--split", type=int, help="verbose output")
    parser.add_argument("--part", type=int, default=1, help="verbose output")
    args = parser.parse_args()
    return args


class BLIP2Preprocessor:
    def __init__(self, processor):
        self.processor = processor

    def __call__(self, image, keywords=None):
        # if keywords is not None:
        #     prompt = f"Question: Generate a description for this image and try to include the terms {', '.join(keywords)}?. Answer:"
        # else:
        #     prompt = f"Question: Generate a description for this image? Answer:"
        keywords = random.choices(keywords, k=5)
        if keywords is not None:
            prompt = f"Create a description of up to three sentences for this image and try to include the terms {', '.join(keywords)}."
        else:
            prompt = f"Create a description of up to three sentences for this image."

        inputs = self.processor(images=image, text=prompt, return_tensors="pt")

        # print(inputs)
        outputs = {
            "pixel_values": inputs["pixel_values"][0],  # .to(torch.float16),
            "input_ids": inputs["input_ids"][0],
            "attention_mask": inputs["attention_mask"][0],
        }

        # if "qformer_input_ids" in inputs:
        outputs["qformer_input_ids"] = inputs["qformer_input_ids"][0]
        # if "qformer_attention_mask" in inputs:
        outputs["qformer_attention_mask"] = inputs["qformer_attention_mask"][0]
        # print(inputs, flush=True)
        # inputs = {k: v[0] for k, v in inputs.items()}

        return outputs


def main():
    args = parse_args()
    # model_name = "Salesforce/blip2-opt-2.7b"
    # model_name = "nielsr/instructblip-vicuna-7b"
    # model_name = "nielsr/instructblip-flan-t5-xxl"
    model = InstructBlipForConditionalGeneration.from_pretrained("nielsr/instructblip-flan-t5-xl")
    processor = InstructBlipProcessor.from_pretrained("nielsr/instructblip-flan-t5-xl")

    print("LOAD MODELLE")
    # processor = AutoProcessor.from_pretrained(model_name)
    # model = Blip2ForConditionalGeneration.from_pretrained(model_name, torch_dtype=torch.float16)

    device = "cuda" if torch.cuda.is_available() else "cpu"
    print("LOAD DATALOADER")
    loader = iconclass_dataloader(
        annotation_path=args.annotation,
        image_path=args.image_path,
        mapping_path=args.mapping_path,
        model_preprocessor=BLIP2Preprocessor(processor),
        part=args.part,
        split=args.split,
        exists_path=args.exists_path,
    )
    model.to(device)

    print("START INFERENCE")
    with open(args.output_path, "w") as f:
        for i, (sample) in enumerate(loader):
            # print(f"{i}/{len(anno)}", flush=True)
            if sample is None:
                continue
            id, inputs = sample

            inputs = {k: v.to(device) for k, v in inputs.items()}
            # generated_ids = model.generate(**inputs, max_newtokens=128)
            print(inputs)
            print(inputs.keys())
            generated_ids = model.generate(
                **inputs,
                do_sample=False,
                num_beams=1,
                max_length=256,
                min_length=1,
                top_p=0.9,
                repetition_penalty=1.5,
                length_penalty=1.0,
                temperature=1,
            )
            generated_text = processor.batch_decode(generated_ids, skip_special_tokens=True)[0].strip()
            print(generated_text)
            descriptions = [generated_text]

            f.write(json.dumps({"id": id[0], "descriptions": descriptions}) + "\n")

    return 0


if __name__ == "__main__":
    sys.exit(main())
```

I call it with this arguments (The files are a bit different then they are in the huggingface and the zonodo files)

```sh
python iconclass_train_blip2_txt.py -a /nfs/data/iart/iconclass/data_202002/train.jsonl --label_path iconclass_kw_en.json --mapping_path /nfs/data/iart/iconclass/data_202002/mapping.jsonl --output_path blip2_7.jsonl  --split 8 --part 7 --image_path /nfs/data/iart/iconclass/data_202002/packed/images/
```

I know the file miss some text, that couldn't generated I think. The dataset file handle this during the training.
